SU2_CFD/include/solvers/CSolver.hpp

```diff
@@ -0,0 +1,4837 @@
+
+  /*!
+   * \brief Pure virtual function, all derived solvers MUST implement a method returning their "nodes".
+   * \note Don't forget to call SetBaseClassPointerToNodes() in the constructor of the derived CSolver.
+   * \return Nodes of the solver, upcast to their base class (CVariable).
+   */
+  virtual CVariable* GetBaseClassPointerToNodes() = 0;
+
+  /*!
+   * \brief Call this method to set "base_nodes" after the "nodes" variable of the derived solver is instantiated.
+   * \note One could set base_nodes directly if it were not private but that could lead to confusion
+   */
+  inline void SetBaseClassPointerToNodes() { base_nodes = GetBaseClassPointerToNodes(); }
+
+private:
+
+  /*--- Private to prevent use by derived solvers, each solver MUST have its own "nodes" member of the
+   most derived type possible, e.g. CEulerVariable has nodes of CEulerVariable* and not CVariable*.
+   This variable is to avoid two virtual functions calls per call i.e. CSolver::GetNodes() returns
+   directly instead of calling GetBaseClassPointerToNodes() or doing something equivalent. ---*/
+  CVariable* base_nodes;  /*!< \brief Pointer to CVariable to allow polymorphic access to solver nodes. */
+
+public:
+
+  CSysVector<su2double> LinSysSol;    /*!< \brief vector to store iterative solution of implicit linear system. */
+  CSysVector<su2double> LinSysRes;    /*!< \brief vector to store iterative residual of implicit linear system. */
+  CSysVector<su2double> LinSysAux;    /*!< \brief vector to store iterative residual of implicit linear system. */
+#ifndef CODI_FORWARD_TYPE
+  CSysMatrix<passivedouble> Jacobian; /*!< \brief Complete sparse Jacobian structure for implicit computations. */
+  CSysSolve<passivedouble>  System;   /*!< \brief Linear solver/smoother. */
+#else
+  CSysMatrix<su2double> Jacobian;
+  CSysSolve<su2double>  System;
+#endif
+
+  CSysMatrix<su2double> StiffMatrix; /*!< \brief Sparse structure for storing the stiffness matrix in Galerkin computations, and grid movement. */
+
+  CSysVector<su2double> OutputVariables;    /*!< \brief vector to store the extra variables to be written. */
+  string* OutputHeadingNames;               /*!< \brief vector of strings to store the headings for the exra variables */
+
+  CVerificationSolution *VerificationSolution; /*!< \brief Verification solution class used within the solver. */
+
+  vector<string> fields;
+  /*!
+   * \brief Constructor of the class.
+   */
+  CSolver(bool mesh_deform_mode = false);
+
+  /*!
+   * \brief Destructor of the class.
+   */
+  virtual ~CSolver(void);
+
+  /*!
+   * \brief Allow outside access to the nodes of the solver, containing conservatives, primitives, etc.
+   * \return Nodes of the solver.
+   */
+  inline CVariable* GetNodes() const {
+    assert(base_nodes!=nullptr && "CSolver::base_nodes was not set properly, see brief for CSolver::SetBaseClassPointerToNodes()");
+    return base_nodes;
+  }
```

Comment:
Methods returning non-const pointers to class members should not be declared const IMO. While the method itself does not change the state of the object, it still provides a way to (changing the nodes in this case), this would allow a function taking a const CSolver* to change its via such methods which would be misleading.
So, until we can return const CVariable*, this should not be const.

Reply:
Fair enough, I agree.